File: app/dtos/pets.py
from datetime import datetime
from typing import Annotated

from pydantic import BaseModel, Field, field_validator

from app.models.pets import PetGrowthStage, PetType


class VirtualPetCreateRequest(BaseModel):
    pet_type: PetType
    pet_name: Annotated[str, Field(min_length=1, max_length=50)]

    @field_validator("pet_name")
    @classmethod
    def validate_pet_name(cls, value: str) -> str:
        stripped = value.strip()
        if not stripped:
            raise ValueError("펫 이름은 1~50자로 입력해주세요.")
        return stripped


class VirtualPetNameUpdateRequest(BaseModel):
    pet_name: Annotated[str, Field(min_length=1, max_length=50)]

    @field_validator("pet_name")
    @classmethod
    def validate_pet_name(cls, value: str) -> str:
        stripped = value.strip()
        if not stripped:
            raise ValueError("펫 이름은 1~50자로 입력해주세요.")
        return stripped


class VirtualPetResponse(BaseModel):
    pet_id: int
    pet_type: PetType
    pet_name: str
    level: int
    experience: int
    next_level_experience: int
    growth_stage: PetGrowthStage
    health_percent: int
    happiness_percent: int


class VirtualPetCreateResponse(BaseModel):
    pet_id: int
    pet_type: PetType
    pet_name: str
    level: int
    experience: int
    growth_stage: PetGrowthStage


class VirtualPetNameUpdateResponse(BaseModel):
    pet_id: int
    pet_name: str


class PetRewardTaskResponse(BaseModel):
    task_type: str
    title: str
    reward_experience: int
    is_completed: bool


class PetRecentActivityResponse(BaseModel):
    activity_type: str
    description: str
    experience_delta: int
    created_at: datetime


class VirtualPetStatusResponse(BaseModel):
    has_pet: bool
    pet: VirtualPetResponse | None = None
    today_tasks: list[PetRewardTaskResponse] = Field(default_factory=list)
    recent_activities: list[PetRecentActivityResponse] = Field(default_factory=list)

File: frontend/src/api/pets.ts
import { apiRequest } from "./client";

export type PetType = "DOG" | "CAT";
export type PetGrowthStage = "STAGE_1" | "STAGE_2" | "STAGE_3";

export type VirtualPet = {
  pet_id: number;
  pet_type: PetType;
  pet_name: string;
  level: number;
  experience: number;
  next_level_experience: number;
  growth_stage: PetGrowthStage;
  health_percent: number;
  happiness_percent: number;
};

export type PetRewardTask = {
  task_type: string;
  title: string;
  reward_experience: number;
  is_completed: boolean;
};

export type PetRecentActivity = {
  activity_type: string;
  description: string;
  experience_delta: number;
  created_at: string;
};

export type VirtualPetStatus = {
  has_pet: boolean;
  pet: VirtualPet | null;
  today_tasks: PetRewardTask[];
  recent_activities: PetRecentActivity[];
};

export type CreateVirtualPetBody = {
  pet_type: PetType;
  pet_name: string;
};

export type VirtualPetCreateResponse = {
  pet_id: number;
  pet_type: PetType;
  pet_name: string;
  level: number;
  experience: number;
  growth_stage: PetGrowthStage;
};

export type VirtualPetNameUpdateResponse = {
  pet_id: number;
  pet_name: string;
};

export type PetRewardClaimResponse = {
  awarded_experience: number;
  claimed_task_count: number;
  level: number;
  experience: number;
  next_level_experience: number;
  growth_stage: PetGrowthStage;
};

export type PetCatalogSummary = {
  total_count: number;
  unlocked_count: number;
  completion_rate: number;
};

export type PetCatalogItem = {
  catalog_id: string;
  pet_type: PetType;
  display_name: string;
  is_unlocked: boolean;
  unlock_condition: string;
  affinity_score: number | null;
};

export type PetCatalog = {
  summary: PetCatalogSummary;
  items: PetCatalogItem[];
};

export async function getMyVirtualPet(token?: string) {
  return apiRequest<{ data: VirtualPetStatus }>("/virtual-pets", { token });
}

export async function createVirtualPet(body: CreateVirtualPetBody, token?: string) {
  return apiRequest<{ data: VirtualPetCreateResponse }>("/virtual-pets", {
    method: "POST",
    body: JSON.stringify(body),
    token,
  });
}

export async function updateVirtualPetName(petName: string, token?: string) {
  return apiRequest<{ data: VirtualPetNameUpdateResponse }>("/virtual-pets/me/name", {
    method: "PATCH",
    body: JSON.stringify({ pet_name: petName }),
    token,
  });
}

export async function claimVirtualPetRewards(token?: string) {
  return apiRequest<{ data: PetRewardClaimResponse }>("/virtual-pets/reward-tasks/claims", {
    method: "POST",
    token,
  });
}

export async function getPetCatalog(petType?: PetType, token?: string) {
  const query = petType ? `?pet_type=${petType}` : "";
  return apiRequest<{ data: PetCatalog }>(`/virtual-pets/catalog${query}`, { token });
}
